### ISSUE TITLE
fmha does not work for padding input

### ISSUE THREAD

**Issue by catcherwill** [NVIDIA/apex#1119]
Hi, i find fmha does not work for padding input. the committed [test-file](https://github.com/NVIDIA/apex/blob/master/apex/contrib/test/fmha/test_fmha.py) use all-1 tensor as mask.
As we known, bert input has padding, so i construct input mask which contains 0 (refer to method: _create_input_mask_) . 
the test code is as follows.

```
import torch
import numpy as np
import unittest
import math
import fmhalib as mha

def py_mha(qkv, amask, b, s, h, d):
    # qkv.shape: [b,s,h,3,d]
    qkv = qkv.view(b, s, h, 3, d)
    q = qkv[:, :, :, 0, :].permute(0,2,1,3)
    k = qkv[:, :, :, 1, :].permute(0,2,1,3)
    v = qkv[:, :, :, 2, :].permute(0,2,1,3)
    p = torch.matmul(q.float(), k.permute(0,1,3,2).float())
    p_masked = p / math.sqrt(d) + (1.0 - amask) * -100000000.0
    s = torch.softmax(p_masked, -1).to(qkv.dtype)
    ctx = torch.matmul(s, v)
    ctx = ctx.permute(0,2,1,3).contiguous()

    ctx.retain_grad()

    return ctx


def create_input_mask(batch, seq_len):
    device = torch.device('cuda')
    mask = torch.zeros(batch, seq_len, dtype=torch.int32, device=device)
    input_lens = torch.randint(1, seq_len, (batch,))
    # input_lens = torch.tensor([seq_len] * batch, dtype=torch.int32, device=device)

    for i in range(batch):
        mask[i, 0:input_lens[i]] = 1
    return mask


def np_assert_allclose(actual, desired, rtol=1e-3, atol=1e-3):
    np.testing.assert_allclose(actual, desired, rtol=rtol, atol=atol, verbose=True)


def to_numpy(tensor):
    return tensor.cpu().detach().numpy()


class TestFMHA(object):

    def run_test(self, s):

        torch.manual_seed(1234)
        torch.cuda.manual_seed(1234)
        
        dtype = torch.float16
        device = torch.device('cuda')

        b = 32  # batch_size
        h = 16  # head_nums
        d = 64  # head_dims

        # create mask, its shape: [batch, seq_len]
        input_mask = create_input_mask(b, s)
        actual_seqlens = input_mask.count_nonzero(dim=1).to(dtype=torch.int32)

        slens = [s] * b  # all sequence length, and then accumulate it
        a = torch.tensor(np.array([0] + slens), dtype=torch.int32)
        cu_seqlens = torch.cumsum(a, 0).to(dtype=torch.int32, device=device)
        input_mask = input_mask.unsqueeze(1).unsqueeze(2)

        qkv = torch.randn((b,s,h,3,d), device=device, dtype=dtype)
        qkv_vs = qkv.permute(0,1,3,2,4).contiguous().view(b*s, 3, h,d)
        qkv.requires_grad = True

        ctx, S_ = mha.fwd(qkv_vs, cu_seqlens, actual_seqlens, 0.0, s, True, None)
        ctx = ctx.view(b,s,h,d)
    
        # ctx_ref = py_mha(qkv, amask, b,s,h,d)
        ctx_ref = py_mha(qkv, input_mask, b, s, h, d)
        np_assert_allclose(to_numpy(ctx_ref.float()), to_numpy(ctx.float()))
    
        labels = torch.randn_like(ctx_ref)
        diff = ctx_ref - labels
        l = (diff * diff).sum() / b
        l.backward()
    
        dw = ctx_ref.grad.permute(0,2,1,3)
        dw2 = dw.permute(0,2,1,3).clone().detach().contiguous()
    
        dqkv2, _ = mha.bwd(dw2, qkv_vs, S_, cu_seqlens, actual_seqlens, 0.0, s)
        dqkv2 = dqkv2.permute(0,2,1,3).view(b,s, h,3,d)

        np_assert_allclose(to_numpy(qkv.grad.float()), to_numpy(dqkv2.float()))
```

@yjk21 thanks for your great work, and could you help me share more about this ? thank you very much.

**Comment by yjk21** [comment]
Hi, can you please try with sequences densely packed to the left?
For that, you can replace `slens = [s] * b` with `input_mask.sum(-1)`, and adjust the size of the output tensor.
Thanks for giving it a try.

**Comment by catcherwill** [comment]
> Hi, can you please try with sequences densely packed to the left?
> For that, you can replace `slens = [s] * b` with `input_mask.sum(-1)`, and adjust the size of the output tensor.
> Thanks for giving it a try.

hi, yjk21, really appreciate for your quick response. and i tried, `input_mask.sum(-1)` and `actual_seqlen = input_mask.count_nonzero(dim=1).to(dtype=torch.int32)` is same.
```
(Pdb) input_mask.sum(-1)
tensor([ 47,  56,  63, 125, 105, 108,  40,  65,   9,  43,  47,  43,  40,  65,
         62, 116,   7,  36,  93,  62,  88,  28, 110,  10,   1,  76,  31, 114,
         24, 121, 123,  41], device='cuda:0')
(Pdb) p actual_seqlen
tensor([ 47,  56,  63, 125, 105, 108,  40,  65,   9,  43,  47,  43,  40,  65,
         62, 116,   7,  36,  93,  62,  88,  28, 110,  10,   1,  76,  31, 114,
         24, 121, 123,  41], device='cuda:0', dtype=torch.int32)
```
did i get you point? could you share a correct test case based on my code if you have bandwith? thanks again.

**Comment by catcherwill** [comment]
Finally, the issue is resolved, when testing fmhalib with padding input, we need to unpad the qkv.
